### PR TITLE
include fasth2

### PR DESCRIPTION
**Pull request by paulhfu**
Hi @wapu , I included my implementation for a "fast" Householder product. You will find the final project report under https://github.com/paulhfu/Faster-Orthogonal-Parameterization-with-Householder-Matrices/blob/master/report/report.pdf 
All my implementations are in https://github.com/paulhfu/Faster-Orthogonal-Parameterization-with-Householder-Matrices/blob/master/FastH/fast_h.py 
Some tests are in https://github.com/paulhfu/Faster-Orthogonal-Parameterization-with-Householder-Matrices/blob/master/test/test_fast_h.py 
I also altered the tests in this repo to include Householder perms. Let me know if I should revert that.

### CODE REVIEW

File: FrEIA/modules/orthogonal.py [hunk 1]
```diff
@@ -82,49 +139,32 @@ def __init__(self, dims_in, dims_c=[], n_reflections=1, fixed=False):
                 # init close to identity
                 init = torch.eye(self.width, self.n_reflections)
                 init += torch.randn_like(init) * 0.1
-            Vs = torch.unbind(init, dim=-1)
-            self.Vs = [nn.Parameter(V) for V in Vs]
-            for i, V in enumerate(self.Vs):
-                V.requires_grad = not self.fixed
-                self.register_parameter(f'V_{i}', V)
+            Vs = init.transpose(-1, -2)
```
Comment (paulhfu) [inline FrEIA/modules/orthogonal.py:142]: Is there a reason why the vectors are stored as tuples? ```_fast_h``` needs ```V``` to be a tensor.

File: FrEIA/modules/orthogonal.py [hunk 2]
```diff
@@ -82,49 +139,32 @@ def __init__(self, dims_in, dims_c=[], n_reflections=1, fixed=False):
...
-                if not rev:
-                    V = Vs[i]
-                else:
-                    V = Vs[-i - 1]
-                VVt = torch.matmul(V.unsqueeze(-1), V.unsqueeze(-2))
-                VtV = torch.matmul(V.unsqueeze(-2), V.unsqueeze(-1)).squeeze(-1)
-                xW = xW - torch.matmul(xW.unsqueeze(-2), VVt).squeeze() * (2/VtV)
-            return [xW]
-
+            Vs = c[0].reshape(-1, self.width, self.n_reflections).transpose(-1, -2)
```
Comment (paulhfu) [inline FrEIA/modules/orthogonal.py:156]: I don't know if I got this right, does ```self.conditional=True``` mean that the ```v```-vectors are explicitly handed to the function? Is it then correct to calculate ```W``` that way and transpose it in the case of  ```rev=True```

File: tests/orthogonal_layer.py [hunk 3]
```diff
@@ -13,10 +13,11 @@
...
 inp = InputNode(inp_size, name='input')
 orthog_layer = Node([inp.out0], orthogonal_layer, {'correction_interval':100})
-permute = Node([orthog_layer.out0], permute_layer, {'seed':0})
-outp= OutputNode([permute.out0], name='output')
+permute_1 = Node([orthog_layer.out0], permute_layer, {'seed':0})
+permute_2 = Node([permute_1.out0], HouseholderPerm, {'n_reflections':10})
+outp= OutputNode([permute_2.out0], name='output')
 
-test_net = ReversibleGraphNet([inp, orthog_layer, permute, outp])
+test_net = ReversibleGraphNet([inp, orthog_layer, permute_1, permute_2, outp])
```
Comment (paulhfu) [inline tests/orthogonal_layer.py:20]: The tests run fine (with commented early returns in ```test_inverse``` and ```test_cuda```)

File: FrEIA/modules/orthogonal.py (hunk 2 above)
Comment (ardizzone) [reply FrEIA/modules/orthogonal.py:156]: Yes, I belive with conditional, the module should be invertible if the condition is kept fix.
So this solution looks right!

File: FrEIA/modules/orthogonal.py (hunk 1 above)
Comment (ardizzone) [reply FrEIA/modules/orthogonal.py:142]: I don't think there was a particular reason.

**Comment by ardizzone** [comment]
Thank you very much Paul, sorry for being slow with the merge!

**Pull request merged**